New Module : aos_ip_pool as part of network/aos

dgarros opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
network/apstra/aos_ip_pool

##### ANSIBLE VERSION
```
ansible 2.3.0 (aos_ip_pool 058e112d07) last updated 2017/02/03 18:14:02 (GMT -700)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
I'm working for [Apstra](http://www.apstra.com/), we are making to automate Datacenter Networks and we are developed a dozen of Ansible module and one dynamic inventory to control our product with Ansible.   
We have tried to follow Ansible's best practices as much as possible, all our modules are idempotent and support the mode `--check`. 

This pull request is the first one, once a got feedback on this one I will open other PR for the other modules.

The module `aos_ip_pool` is used to manage IP Pool on our solutions.  

Looking forward to your feedback

dgarros committed:
Initial version of aos_ip_pool module

dgarros committed:
Add examples for IP Pool

dgarros committed:
Break down ip_pool into smaller function

dgarros committed:
Refactor do_load_resource

dgarros committed:
Add get_display_name_from_file

dgarros committed:
Add ‘src’ as an option to load ip pool from JSON file

dgarros committed:
Rename directory network/apstra to network/aos

dgarros committed:
Remove exception handling temporary

dgarros committed:
Remove all ‘Exception as XX’ to be python 2.4 compatible

dgarros committed:
Replace ‘== False’ with ‘is False’ for PEP8 Test

dgarros committed:
Update documentation to be Yaml compatible

gundalow commented on lib/ansible/module_utils/aos.py:
```diff
@@ -0,0 +1,132 @@
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
```
`module_utils` should be BSD. You can copy the text from another library in this module

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+author: damien@apstra.com (@dgarros)
+version_added: "2.3"
+short_description: Manage AOS IP Pool
+description:
```
All `description:` (apart from `short_description:`) must be full sentences, ie start with a capital letter and end with a full stop.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+    required: true
+  name:
+    description:
+      - Name of the IP Pool to manage
```
Name of the IP Pool to manage. Only one of I(name), I(id) or I(src) can be set.

This apply to a few option descriptions.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+    required: false
+  state:
+    description:
+      - Indicate
```
A little more detail here would be good.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+  subnets:
+    description:
+      - Filepath to JSON file containing the collection item data
+    default: []
```
You can delete this line.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+EXAMPLES = '''
+# Create an IP Pool with one subnet
+
+    - name: "Create IP Pool"
```
- name: "Create an IP Pool with one subnet"

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+'''
+
+EXAMPLES = '''
+# Create an IP Pool with one subnet
```
Delete the comments, and set them as `- name: ` instead

Then fix the indentation `- name: ` shouldn't have any spaces before it. This is to allow people to copy & paste the examples straight into their playbooks.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+import json
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.aos import *
```
Import only what you need and not `*`

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+
+    margs = module.params
+
+    # TODO add Exception handling for login error
```
TODO add Exception handling for login error

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,310 @@
+options:
+  session:
+    description:
+      - An existing AOS session as obtained by aos_login module
```
Once aos_login has been merged you can change this to `M(aos_login)` to generate a hyperlink to the Module docs for aos_login.

dgarros committed:
Lisg all method imported from module_utils.aos

dgarros committed:
Refactor to align with collection.find() changes

dgarros committed:
Update examples by @gundalow’s  recommendations

dgarros committed:
Update Documentation per @gundalow’s recommendations

dgarros committed:
Change the license per @gundalow recommendation

dgarros committed:
Add exception handling for get_aos_session

dgarros commented:
@gundalow Thanks for reviewing this module, appreciate your feedback.
 I made all changes you requested. Please let me know if you see more things that needs to be change.

gundalow reviewed (approved):
Looks good, thanks

privateip commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,298 @@
+            session=dict(required=True, type="dict"),
+            name=dict(required=False ),
+            id=dict(required=False ),
+            src=dict(required=False),
```
should be `type='path'`

privateip commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,298 @@
+    item_id = False
+
+    if margs['src'] is not None:
+        item_name = get_display_name_from_file(module, margs['src'])
```
This will fail if the playbook task uses `delegate_to`.   Consider instead of accepting the file contents instead of a source file name

dgarros replied:
I'm not familiar with `delegate_to`, would you have an example that is doing what you propose ?
thks

privateip commented on lib/ansible/module_utils/aos.py:
```diff
@@ -0,0 +1,142 @@
+    except KeyError:
+        module.fail_json(msg="src data missing display_name, check file contents")
+
+    except SessionRqstError:
```
I think this will fail a `NameError`

dgarros committed:
Change Auth format and add check_aos_version() to check minimum version

dgarros committed:
Add a check for minimum version

dgarros committed:
Refactor ‘src’ into ‘content’ to allow more options
Remove get_display_name_from_file in aos.py
Add content_to_dict in aos.py

dgarros commented:
@privateip, I made few changes to this module following your review, the main one is that I replaced `src` with `content` to allow more options. 
 It's now possible to pass a JSON string, a Yaml string or a variable directly. I also added a parameter `content_format` to indicate in which format the content is provided.

```
- name: "Load IP Pool from a JSON file"
  aos_ip_pool:
    session: "{{ session_ok }}"
    content: "{{ lookup('file', 'resources/ip_pool_saved.json') }}"
    state: present

- name: "Load IP Pool from a YAML file"
  aos_ip_pool:
    session: "{{ session_ok }}"
    content: "{{ lookup('file', 'resources/ip_pool_saved.yaml') }}"
    content_format: yaml
    state: present

- name: "Load IP Pool from a Variable"
  aos_ip_pool:
    session: "{{ session_ok }}"
    content:
      display_name: my-ip-pool
      id: 4276738d-6f86-4034-9656-4bff94a34ea7
      subnets:
        - network: 172.10.0.0/16
        - network: 172.12.0.0/16
    content_format: var
    state: present
```

I also made couple of minor changes: 
- Add a minimum version check for the library aos-pyez
- Change the authentication format to match with latest version of aos-pyez
- I made the exception handling more generic in `do_load_resource` following your comment


Please let me know if you see something else that needs to be changed.
thanks

dgarros commented:
I see some errors in shippable related to `test/units/modules/network/vyos/`, not sure how these arrived here.
I don't think the errors reported are related to the aos module

dgarros committed:
Fix variable name in do_load_resource

dgarros commented:
I rebased from `devel` and tests are passing now

bcoca commented on lib/ansible/module_utils/aos.py:
```diff
@@ -0,0 +1,174 @@
+
+"""
+import json
+import yaml
```
yaml is not required on target machines, even if this module is expected to run with `connection: local` you might be run on 'delegated' machine which does not have Ansible installed,

you should use same `try/HAS_=true` pattern used below

bcoca commented:
aside from the initial connection there is no exception handling, it is a bug if a module returns an exception instead of a JSON document.

dgarros committed:
Add mention of aos.py in module_utilities doc

dgarros committed:
Add try/except for import yaml

dgarros committed:
Add try/Except around main block of code and function

dgarros commented:
@bcoca, thanks for your feedback. I made some changes to address your comments.
I added a `try/except` block in several places on the main function or code block.

Please let me know if you have more comments.

@privateip, did my changes addressed your comments as well ? I refactor the parameter `src` to be more friendly with `deletegate_to`

Also I added a mention of `module_utils/aos.py` into the documentation as suggested.

Looking forward to your inputs.

bcoca commented:
@dgarros the yaml import is still an issue

for content, FYI, JSON is a subset of YAML so yaml can normally parse both transparently. Also you can use detection (see our type=json implementation) to allow all 3 formats transparently w/o need of a content_format option.

alikins commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,355 @@
+
+    return subnets
+
+def create_new_ip_pool(ip_pool, name, subnets=[]):
```
lib/ansible/modules/network/aos/aos_ip_pool.py:191:1: M511 - mutable default arg of type List

There is only one caller of create_new_ip_pool() and it provides all three args, so subnets doesnt need to be keyword arg or optional.

alikins commented on lib/ansible/module_utils/aos.py:
```diff
@@ -0,0 +1,184 @@
+
+try:
+    from apstra.aosom.session import Session
+    import apstra.aosom.exc as aosExc
```
This import doesn't appear to be used.

alikins commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,355 @@
+    in_use = False
+    if my_pool.value:
+        if my_pool.value['status'] != 'not_in_use':
+            in_use = True
```
As far as I can tell, the 'in_use' is never used.

dgarros committed:
Refactor to auto detect content_format, update doc accordingly

dgarros committed:
Change create_new_ip_pool inputs

dgarros committed:
Remove unused import

dgarros committed:
Remove in_use as it’s never used

dgarros commented:
@bcoca, I removed `content_format` as suggested and i updated the documentation accordingly. Now the format of the input is automatically detected. 
Please let me know if you see something else

@alikins thanks for your feedback, I made the changes to address your comments. Very good catches. Please let me know if you see something else.

dgarros committed:
Fix doc format

dgarros commented:
I pushed some wrong commits in my branch, this is why @ansibot  requested a rebase.
I removed cleaned up everything, sorry for the noise.

gundalow commented on lib/ansible/modules/network/aos/aos_ip_pool.py:
```diff
@@ -0,0 +1,344 @@
+    create and delete IP Pool by Name, ID or by using a JSON File. This module
+    is idempotent and support the I(check) mode. It's using the AOS REST API
+requirements:
+  - aos-pyez
```
I'm not sure how old aos-pyez 0.6.0 is, though you may wish to detail that here.
`- "aos-pyez >= 0.6.0"`

dgarros committed:
Add version number in requirement doc

dgarros commented:
@gundalow Very good point, I made the change 

@privateip @alikins Is there anything else you would like to change on this PR ? 
it would be great to get this one merged. I have more PR to open for additional modules before the 2.3 deadline next week
Please let me know if you have more comments or questions

gundalow commented:
Merged, thanks for the prompt updates, the rest should go faster. 

So you are now free to raise 5 PRs, each containing a single module. Feel free to `@gundalow` in the PRs so I'll get notified.

dgarros commented:
thanks a lot for your help
